APEXMALHAR-2085: Operator supporting the Beam concepts of windowing, watermarks, triggering and accumulation

davidyan74 opened the pull request:
This review-only PR contains the interfaces and a first rough draft of the implementation of the operator that supports the concepts of windowing, watermarks, triggering and accumulation mode specified by the Apache Beam API. 

When you review this PR, please:
- note the TODO comments
- look at how the pi and wordcount samples work
- try running the pi and wordcount samples (there is a main method that you can run in your IDE)

thank you.

tweise commented on stream/src/main/java/org/apache/apex/malhar/stream/window/Accumulation.java:
```diff
@@ -0,0 +1,73 @@
+ *
+ */
+@InterfaceStability.Evolving
+public interface Accumulation<InputT, AccumT, OutputT>
```
Might be helpful to add an example where AccumT is different from OutputT

davidyan74 replied:
done

tweise commented on stream/src/main/java/org/apache/apex/malhar/stream/window/WindowedStorage.java:
```diff
@@ -0,0 +1,134 @@
+ * This interface is for a key/value store for storing data for windowed streams.
+ * The key to this key/value store is a pair of (Window, K)
+ *
+ * Note that this interface does not take recovery into consideration.
```
Maybe state that implementations need to do that and how.

tweise commented on stream/src/main/java/org/apache/apex/malhar/stream/window/WindowedStorage.java:
```diff
@@ -0,0 +1,134 @@
+   *
+   * @param timestamp
+   */
+  Set<Window> windowsIncluding(long timestamp);
```
Any insight on how these operations can be implemented efficiently with the backing storage for high throughput streams?

tweise commented on stream/src/main/java/org/apache/apex/malhar/stream/window/impl/DefaultWindowedStorageImpl.java:
```diff
@@ -0,0 +1,120 @@
+/**
+ * Created by david on 6/14/16.
+ */
+public class DefaultWindowedStorageImpl<K, V> implements WindowedStorage<K, V>
```
In the spirit of designing an interface that allows for efficient storage operation with large state, can you consider 
https://issues.apache.org/jira/browse/APEXMALHAR-2026
as well as managed state?

tweise commented on stream/src/main/java/org/apache/apex/malhar/stream/window/impl/WindowedOperatorImpl.java:
```diff
@@ -0,0 +1,436 @@
+  private WindowedStorage<KeyT, AccumT> dataStorage;
+  private WindowedStorage<KeyT, AccumT> retractionStorage;
+
+  private TreeMap<Window, WindowState> windowStateMap = new TreeMap<>();
```
Assumes that there will only be a few windows?

davidyan74 replied:
I have abstracted out the storage for the window state

tweise commented on stream/src/main/java/org/apache/apex/malhar/stream/window/impl/WindowedOperatorImpl.java:
```diff
@@ -0,0 +1,436 @@
+    }
+  }
+
+  private List<Window.TimeWindow> getTimeWindowsFromTimestamp(long timestamp)
```
ForTimestamp? Even though it's private, some javadoc wouldn't hurt here.

davidyan74 replied:
done

tweise commented on stream/src/main/java/org/apache/apex/malhar/stream/window/impl/WindowedOperatorImpl.java:
```diff
@@ -0,0 +1,436 @@
+  public void setup(Context.OperatorContext context)
+  {
+    if (this.firstWindowMillis <= 0) {
+      this.firstWindowMillis = System.currentTimeMillis();
```
Should eventually be the DAG start time. We need to add that in core in the future. Till then we could also derive it from first windowId.

tushargosavi commented on stream/src/main/java/org/apache/apex/malhar/stream/window/WindowedStorage.java:
```diff
@@ -0,0 +1,134 @@
+   * @param window
+   * @return
+   */
+  Set<Map.Entry<K, V>> entrySet(Window window);
```
Can it be Iterator<Map.Entry<K,V>>  In case state is huge and stored in fs

davidyan74 replied:
done

tushargosavi commented on stream/src/main/java/org/apache/apex/malhar/stream/api/WindowedStream.java:
```diff
@@ -53,4 +60,121 @@
+   * @param reduce reduce function
+   * @return new stream of same type
+   */
+  <STREAM extends ApexStream<T>> STREAM reduce(Function.ReduceFunction<T> reduce);
```
reduceByKey ??

tweise commented on stream/pom.xml:
```diff
@@ -94,6 +94,18 @@
       <version>3.2.1</version>
     </dependency>
+    <dependency>
+      <!-- required by twitter demo -->
```
Why this dependency?

tweise commented on stream/src/main/java/org/apache/apex/malhar/stream/window/Tuple.java:
```diff
@@ -0,0 +1,157 @@
+   *
+   * @param <T>
+   */
+  class WatermarkTuple<T> extends TimestampedTuple<T> implements Watermark
```
This won't work. Watermarks need to be sent through a different port and stream to handle partitioning. In the future, this would be a prime candidate for custom control tuple.

davidyan74 replied:
Should we add the support for custom control tuple in the core engine sooner to avoid having a separate port for watermark? This will also help with batch support.

Commit:
APEXMALHAR-2085 windowed operator interfaces and implementation

addressing PR comments

Added trigger unit tests

Fixed sliding window bug, added unit tests

Fixed session window bug; added more unit tests

Gives window a chance to trigger before purging because of lateness

added more unit tests

Process watermark only at end window

Renamed WatermarkOpt to Type, and fixed bug when window in retraction storage is not dropped when it's too late

added copyright

rat check

changed WindowOption from an abstract class to an interface

Use mutable objects for accumulated types

Retraction storage needs to be based on the output, not accumulated type

Added more unit tests

added support of fixed lateness in case watermark is not available from upstream

changed name from fixed lateness to fixed watermark

support inheriting windowed tuple from upstream when window option is not given

chinmaykolhatkar commented on library/src/main/java/org/apache/apex/malhar/lib/window/impl/AbstractWindowedOperator.java:
```diff
@@ -0,0 +1,484 @@
+ * values are stored in the storage.
+ *
+ */
+public abstract class AbstractWindowedOperator<InputT, AccumT, OutputT, DataStorageT extends WindowedStorage, AccumulationT extends Accumulation>
```
I don't see AccumT used anywhere in this class. If unused can we remove this from Generic type list?
Same for WindowedOperator too.. AccumT & OutputT is unused.

davidyan74 replied:
Good point. done.

chinmaykolhatkar commented on library/src/main/java/org/apache/apex/malhar/lib/window/impl/KeyedWindowedOperatorImpl.java:
```diff
@@ -0,0 +1,159 @@
+ * This is an implementation of WindowedOperator that takes in key value pairs as input and gives out key value pairs
+ * as output. If your operation is not key based, please use {@link WindowedOperatorImpl}.
+ */
+public class KeyedWindowedOperatorImpl<KeyT, InputValT, AccumT, OutputValT>
```
Can you make this evolving?

davidyan74 replied:
done

chinmaykolhatkar commented on library/src/main/java/org/apache/apex/malhar/lib/window/impl/AbstractWindowedOperator.java:
```diff
@@ -0,0 +1,486 @@
+   */
+  public void processTuple(Tuple<InputT> tuple)
+  {
+    long timestamp = extractTimestamp(tuple);
```
extractTimestamp method is called here, then in getWindowsValue and also in assignWindow for the same tuple.
If its a TimestampedTuple its fine as as its simple a getter method but if TimeExtractor is set the operation of extracting timestamp might get costly depending on the logic in extractor.

Can this call be made only once?

chinmaykolhatkar commented on library/src/main/java/org/apache/apex/malhar/lib/window/impl/AbstractWindowedOperator.java:
```diff
@@ -0,0 +1,486 @@
+  }
+
+  private void assignWindows(List<Window> windows, Tuple<InputT> inputTuple)
+  {
```
Minor nit: Considering WindowedTuple has a addWindow method, maybe it just make sense to have only single parameter to this method "WindowedTuple<InputT> tuple".. Rest all can be derived from this object.

brightchen commented on library/src/main/java/org/apache/apex/malhar/lib/window/Window.java:
```diff
@@ -0,0 +1,179 @@
+ * This interface describes the individual window.
+ */
+@InterfaceStability.Evolving
+public interface Window
```
I think for Window, we could only provide one interface isWithinWIndow(long time); So for GlobalWindow, we can simplly return true. And in SessionWindow, instead of using static method to merge the window, we can just extendWindow(). The Window interface looks like assume there has a fixed boundary(it could be the interface of FixedWindow). Both GlobalWindow and SessionWindow don't have fixed boundary.

davidyan74 replied:
Session window can be extended, and two session windows can be merged into one.

chinmaykolhatkar commented on library/src/main/java/org/apache/apex/malhar/lib/window/impl/AbstractWindowedOperator.java:
```diff
@@ -0,0 +1,490 @@
+    long horizon = currentWatermark - allowedLatenessMillis;
+    if (allowedLatenessMillis >= 0) {
+      // purge window that are too late to accept any more input
+      dataStorage.removeUpTo(horizon);
```
Before removing from storage, should we check whether triggers are fired and data is send to downstream for window that is too late?

davidyan74 replied:
good point, will make the changes

chinmaykolhatkar commented on library/src/main/java/org/apache/apex/malhar/lib/window/impl/AbstractWindowedOperator.java:
```diff
@@ -0,0 +1,490 @@
+        // watermark has not arrived for this window before, marking this window late
+        if (windowState.watermarkArrivalTime == -1) {
+          windowState.watermarkArrivalTime = currentDerivedTimestamp;
+          if (triggerAtWatermark) {
```
Shouldn't triggerAtWatermark be a default option?
As per beam specification, trigger at watermark always happens.

davidyan74 replied:
This gives an option to not trigger at watermark, which may be useful.

chinmaykolhatkar replied:
Can we make this property default true in that case?
If we're going by mean specifications, then even if TriggerOption list not having ON_TIME is set, should make sure by default that trigger happens at watermark.

chinmaykolhatkar commented on library/src/main/java/org/apache/apex/malhar/lib/window/TriggerOption.java:
```diff
@@ -0,0 +1,278 @@
+   */
+  public enum WatermarkOpt
+  {
+    EARLY,
```
The terminology of WatermarkOpt is confusion. can we change this to TriggerType?

davidyan74 replied:
We can. @siyuanh any problem with that?

bhupeshchawda commented on library/src/main/java/org/apache/apex/malhar/lib/window/WindowedStorage.java:
```diff
@@ -0,0 +1,88 @@
+ *
+ * @param <T> The type of the data that is stored per window
+ *
+ * TODO: Look at the possibility of integrating spillable data structure: https://issues.apache.org/jira/browse/APEXMALHAR-2026
```
This seems very similar to the managed state implementations. Can we also look at managed state directly before looking at Spillable data structures?

davidyan74 replied:
@bhupeshchawda We will for sure look at ManagedState.

vrozov commented on library/src/main/java/org/apache/apex/malhar/lib/window/Accumulation.java:
```diff
@@ -0,0 +1,76 @@
+   * @param input
+   * @return
+   */
+  AccumT accumulate(AccumT accumulatedValue, InputT input);
```
What is the reason to make the interface stateless? Should it be stateful? My concern is with the interface promoting boxing/unboxing pattern and leading to an excessive GC.

davidyan74 replied:
@vrozov Good point. I'm assuming you're asking why not: 

``` java
  void accumulate(AccumT accumulatedValue, InputT input);
```

with accumulatedValue updated in place. But doing it will make it a lot less flexible because the underlying storage might not support this kind of operation. For example, if the storage supports get(key) and put(key, value) with get(key) returning not a reference to the actual object (possibly as a result of deserialization), then it would not work.

vrozov replied:
My suggestion is to delegate to a class that implements Accumulation interface handling of accumulation and change interface to 

```
void accumulate(InputT input);
void merge(Accumulation accumulatedValue);
OutputT getOutput();
```

The implementation class will need to define how it handles accumulation and how AccumT is defined. The implementation may use Collection for AccumT or may use primitive types such as int, long or double to accumulate values.

vrozov commented on library/src/test/java/org/apache/apex/malhar/lib/window/SumAccumulation.java:
```diff
@@ -0,0 +1,55 @@
+  @Override
+  public Long accumulate(Long accumulatedValue, Long input)
+  {
+    return accumulatedValue + input;
```
This is an example of the interface promoting an implementation that is subject to boxing and GC. It will be much more efficient to accumulate values in an internal long member variable.
